### DISCUSSION TITLE
Will unison make use of WASM as a backend?

### DISCUSSION THREAD

**Discussion by He-Pin** [unisonweb/unison#5516] [General]
Will unison make use of WASM as a backend?

**Comment by hxrts** [comment]
Also curious about this!

**Comment by rlmark** [comment]
Thanks for asking! Right now, we haven't prioritized WASM as a compilation target for Unison for a couple of reasons, but if you feel strongly that this would allow you to use Unison at your organization or job, be sure to drop a note in the comment thread for us to be aware of. 

* WASM doesn't currently support Unison's model for effects, Abilities, which are based on continuations. There's an open discussion about [adding continuation support to WASM here](https://github.com/WebAssembly/design/issues/1359) from the project maintainers. It's a great read if you're curious about effect handlers and what problems they solve for programming languages. As it stands, performing basic effectful operations in Unison would require some underlying support for continuations in WASM. 
  - There's some interesting research in adding continuations: https://dl.acm.org/doi/10.1145/3426422.3426978, but we're hoping to see the research tested in practice before writing an interpreter. 
* Garbage collection support in WASM is in active development. [Here's the wasm gc MVP doc, if you're curious.](https://github.com/WebAssembly/gc/blob/main/proposals/gc/MVP.md)

Once features like garbage collection and continuation support are more battle-tested, we could consider investing more time into a WASM backend. For now, the folks working on the compiler / interpreter have been focusing their time on performance improvements for the existing Unison runtime, and the JIT compiler.

**Reply by He-Pin** [reply]
Thanks for the detailed input, Unison is cool, I'm asking because if I could have a Unison program running inside my web browser but connected to the Unison Cloud that would be pretty cool too.

**Reply by zacharywhitley** [reply]
I'm not sure about continuations but the GC proposal is complete and supported in Chrome and Wasmtime 27.0 https://bytecodealliance.org/articles/wasmtime-27.0

**Reply by zacharywhitley** [reply]
Tail calls and GC are now an official part of WebAssembly 3.0

**Comment by hxrts** [comment]
I'm actually more interested in compilation to RISC-V, but there was already an outstanding issue for WASM.

**Comment by martinnester** [comment]
There is now [Warp](https://share.unison-lang.org/@dfreeman/warp) available on unison share that can compile Unison code to WASM.

**Reply by zacharywhitley** [reply]
Oh, hell ya! and it's not even my birthday today. Thanks!
